src/Periphery/Permit2Proxy.sol

```diff
@@ -0,0 +1,356 @@
+        );
+
+        // maxApprove token to diamond if current allowance is insufficient
+        LibAsset.maxApproveERC20(
+            IERC20(_permit.permitted.token),
+            LIFI_DIAMOND,
+            _permit.permitted.amount
+        );
+
+        _executeCalldata(_diamondCalldata);
+    }
+
+    /// @notice utitlity method for constructing a valid Permit2 message hash
+    /// @param _diamondCalldata the calldata to execute
+    /// @param _assetId the address of the token to approve
+    /// @param _amount amount of tokens to approve
+    /// @param _nonce the nonce to use
+    /// @param _deadline the expiration deadline
+    function getPermit2MsgHash(
+        bytes calldata _diamondCalldata,
+        address _assetId,
+        uint256 _amount,
+        uint256 _nonce,
+        uint256 _deadline
+    ) external view returns (bytes32 msgHash) {
+        // Token Permissions
+        ISignatureTransfer.TokenPermissions
+            memory tokenPermissions = ISignatureTransfer.TokenPermissions(
+                _assetId,
+                _amount
+            );
+        bytes32 permit = _getTokenPermissionsHash(tokenPermissions);
+
+        // Witness
+        Permit2Proxy.LiFiCall memory lifiCall = LiFiCall(
+            LIFI_DIAMOND,
+            keccak256(_diamondCalldata)
+        );
+        bytes32 witness = _getWitnessHash(lifiCall);
+
+        // PermitTransferWithWitness
+        msgHash = _getPermitWitnessTransferFromHash(
+            PERMIT2.DOMAIN_SEPARATOR(),
+            permit,
+            address(this),
+            _nonce,
+            _deadline,
+            witness
+        );
+    }
+
+    /// Internal Functions ///
+
+    function _getTokenPermissionsHash(
+        ISignatureTransfer.TokenPermissions memory tokenPermissions
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    PermitHash._TOKEN_PERMISSIONS_TYPEHASH,
+                    tokenPermissions.token,
+                    tokenPermissions.amount
+                )
+            );
+    }
+
+    function _getWitnessHash(
+        Permit2Proxy.LiFiCall memory lifiCall
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encode(WITNESS_TYPEHASH, lifiCall));
+    }
+
+    function _getPermitWitnessTransferFromHash(
+        bytes32 domainSeparator,
+        bytes32 permit,
+        address spender,
+        uint256 nonce,
+        uint256 deadline,
+        bytes32 witness
+    ) internal view returns (bytes32) {
+        bytes32 dataHash = keccak256(
+            abi.encode(
+                PERMIT_WITH_WITNESS_TYPEHASH,
+                permit,
+                spender,
+                nonce,
+                deadline,
+                witness
+            )
+        );
+
+        return
+            keccak256(abi.encodePacked("\x19\x01", domainSeparator, dataHash));
+    }
+
+    function _executeCalldata(bytes memory diamondCalldata) internal {
```

Comment:
just wondering...would it make sense that we return the result (`data`) in success case?
Just thinking that for example we do a GenericSwap which in the future would (as we discussed yesterday) return the amount that was swapped. If we call the diamond using Permit2, then this information gets lost on the way. 
But if we were to return whatever comes back from the call, we could decode the bytes to the uint amount in that case. 

Does that make sense? What do you think?